psl-core/src/main/java/org/linqs/psl/model/rule/logical/AbstractLogicalRule.java

```diff
@@ -63,10 +63,10 @@
 
     protected Formula formula;
     protected final DNFClause negatedDNF;
-    private int hash;
 
     public AbstractLogicalRule(Formula formula, String name) {
-        super(name);
+        // Hashcode is set to zero here and then set using the negatedDNF in this constructor.
+        super(name, 0);
```

Comment:
What about something like `super(name, init());`?
And just moving over all the initialization code?
Then I think we can also get rid of `AbstractRule.setHashcode()` and make `AbstractRule.hashcode` final.

Reply:
I removed the setHashCode() method but since we can't call instance methods before calling super I moved all of the initialization into the AbstractLogicalRule constructor.